# Visualization

The first step is to visualize the data table contained in "P_BIOPRO.xpt". The first five observations are shown below.

In [36]:
import pandas as pd

df_p_biopro = pd.read_sas("P_BIOPRO.xpt")

df_p_biopro.head()

,SEQN,LBXSATSI,LBDSATLC,LBXSAL,LBDSALSI,LBXSAPSI,LBXSASSI,LBXSC3SI,LBXSBU,LBDSBUSI,...,LBXSCA,LBDSCASI,LBXSCH,LBDSCHSI,LBXSTP,LBDSTPSI,LBXSTR,LBDSTRSI,LBXSUA,LBDSUASI
0,109264.0,8.0,5.397605e-79,4.3,43.0,172.0,18.0,26.0,9.0,3.21,...,9.4,2.350,170.0,4.396,6.9,69.0,54.0,0.610,3.5,208.2
1,109266.0,15.0,5.397605e-79,3.8,38.0,74.0,14.0,25.0,8.0,2.86,...,8.8,2.200,199.0,5.146,6.8,68.0,86.0,0.971,4.9,291.5
2,109271.0,8.0,5.397605e-79,3.8,38.0,88.0,14.0,30.0,8.0,2.86,...,8.9,2.225,148.0,3.827,7.3,73.0,101.0,1.140,5.0,297.4
3,109273.0,35.0,5.397605e-79,4.6,46.0,70.0,27.0,27.0,11.0,3.93,...,9.7,2.425,168.0,4.344,7.4,74.0,178.0,2.010,5.2,309.3
4,109274.0,19.0,5.397605e-79,4.0,40.0,49.0,20.0,28.0,11.0,3.93,...,9.1,2.275,105.0,2.715,6.8,68.0,151.0,1.705,5.1,303.3


Each of its attributes is displayed, along with the number of rows and columns that comprise the dataset:

In [37]:
df_p_biopro.columns

Index(['SEQN', 'LBXSATSI', 'LBDSATLC', 'LBXSAL', 'LBDSALSI', 'LBXSAPSI',
       'LBXSASSI', 'LBXSC3SI', 'LBXSBU', 'LBDSBUSI', 'LBXSCLSI', 'LBXSCK',
       'LBXSCR', 'LBDSCRSI', 'LBXSGB', 'LBDSGBSI', 'LBXSGL', 'LBDSGLSI',
       'LBXSGTSI', 'LBDSGTLC', 'LBXSIR', 'LBDSIRSI', 'LBXSLDSI', 'LBXSOSSI',
       'LBXSPH', 'LBDSPHSI', 'LBXSKSI', 'LBXSNASI', 'LBXSTB', 'LBDSTBSI',
       'LBDSTBLC', 'LBXSCA', 'LBDSCASI', 'LBXSCH', 'LBDSCHSI', 'LBXSTP',
       'LBDSTPSI', 'LBXSTR', 'LBDSTRSI', 'LBXSUA', 'LBDSUASI'],
      dtype='str')

In [38]:
df_p_biopro.shape

(10409, 41)

The nomenclature used in the P_BIOPRO.xpt file is detailed below:

(This step is not automated and therefore applies only to the P_BIOPRO.xpt file)


<ol>
    <li><b>SEQN</b>: Respondent Sequence Number</li>
    <li><b>LBXSATSI</b>: Alanine Aminotransferase (ALT) (U/L)</li>
    <li><b>LBDSATLC</b>: Alanine Aminotransferase (ALT) Comment Code</li>
    <li><b>LBXSAL</b>: Albumin, refrigerated serum (g/dL)</li>
    <li><b>LBDSALSI</b>: Albumin, refrigerated serum(g/L)</li>
    <li><b>LBXSAPSI</b>: Alkaline Phosphatase (ALP) (IU/L)</li>
    <li><b>LBXSASSI</b>: Aspartate Aminotransferase (AST) (U/L)</li>
    <li><b>LBXSC3SI</b>: Bicarbonate (mmol/L)</li>
    <li><b>LBXSBU</b>: Blood Urea Nitrogen (mg/dL)</li>
    <li><b>LBDSBUSI</b>: Blood Urea Nitrogen (mmol/L)</li>
    <li><b>LBXSCLSI</b>: Chloride (mmol/L)</li>
    <li><b>LBXSCK</b>: Creatine Phosphokinase (CPK) (IU/L)</li>
    <li><b>LBXSCR</b>: Creatinine, refrigerated serum (mg/dL)</li>
    <li><b>LBDSCRSI</b>: Creatinine, refrigerated serum (umol/L)</li>
    <li><b>LBXSGB</b>: Globulin (g/dL)</li>
    <li><b>LBDSGBSI </b>: Globulin (g/L)</li>
    <li><b>LBXSGL</b>: Glucose, refrigerated serum (mg/dL)</li>
    <li><b>LBDSGLSI</b>: Glucose, refrigerated serum (mmol/L)</li>
    <li><b>LBXSGTSI</b>: Gamma Glutamyl Transferase (GGT) (IU/L)</li>
    <li><b>LBDSGTLC</b>: Gamma Glutamyl Transferase (GGT) Comment Code</li>
    <li><b>LBXSIR</b>: Iron, refrigerated serum (ug/dL)</li>
    <li><b>LBDSIRSI</b>: Iron, refrigerated serum (umol/L)</li>
    <li><b>LBXSLDSI</b>: Lactate Dehydrogenase (LDH) (IU/L)</li>
    <li><b>LBXSOSSI</b>: Osmolality (mmol/Kg)</li>
    <li><b>LBXSPH</b>: Phosphorus (mg/dL)</li>
    <li><b>LBDSPHSI</b>: Phosphorus (mmol/L)</li>
    <li><b>LBXSKSI</b>: Potassium (mmol/L)</li>
    <li><b>LBXSNASI</b>: Sodium (mmol/L)</li>
    <li><b>LBXSTB</b>: Total Bilirubin (mg/dL)</li>
    <li><b>LBDSTBSI</b>: Total Bilirubin (umol/L)</li>
    <li><b>LBDSTBLC</b>: Total Bilirubin Comment Code</li>
    <li><b>LBXSCA</b>: Total Calcium (mg/dL)</li>
    <li><b>LBDSCASI</b>: Total Calcium (mmol/L)</li>
    <li><b>LBXSCH</b>: Cholesterol, refrigerated serum (mg/dL)</li>
    <li><b>LBDSCHSI</b>: Cholesterol, refrigerated serum (mmol/L)</li>
    <li><b>LBXSTP</b>: Total Protein (g/dL)</li>
    <li><b>LBDSTPSI</b>: Total Protein (g/L)</li>
    <li><b>LBXSTR</b>: Triglycerides, refrig serum (mg/dL)</li>
    <li><b>LBDSTRSI</b>: Triglycerides, refrig serum (mmol/L)</li>
    <li><b>LBXSUA</b>: Uric acid (mg/dL)</li>
    <li><b>LBDSUASI</b>: Uric acid (umol/L)</li>
</ol>

# Data Cleaning

In the "P_BIOPRO.xpt" dataset, there are columns with repeated attributes but different measurement units. This variation occurs by factors of 100 or 1000, indicating that the values may be expressed in "mg/dL" or "mmol/L". This only affects the numerical representation of the values and does not alter the analysis itself.

The dataset will be modified according to the following criteria:

<ol>
    <li>Columns containing only null values are removed.</li>
    <li>Columns with the same attributes but different measurement units are removed, retaining those expressed in SI units.</li>
    <li>Rows containing only the SEQN value are removed.</li>
</ol>

This approach enables a more effective data cleaning process and improves the overall quality of the dataset.

In [39]:
import pandas as pd

# 1. Load the .xpt file
input_file = 'P_BIOPRO.xpt'
df = pd.read_sas(input_file)

# --- STEP A: REMOVE SPECIFIC COLUMNS ---
columns_to_remove = [
    'LBXSAL', 'LBXSBU', 'LBXSCR', 'LBXSGB', 'LBXSGL', 'LBXSIR', 'LBXSPH', 'LBXSTB', 'LBXSCA', 'LBXSCH',
    'LBXSTP', 'LBXSTR', 'LBXSUA', 'LBDSATLC', 'LBDSTBLC', 'LBDSGTLC'
]
clean_df = df.drop(columns=columns_to_remove, errors='ignore')

# --- STEP A.1: REMOVE COLUMNS IF THE ENTIRE COLUMN IS NULL ---
# axis=1 indicates columns, how='all' indicates that ALL values must be NaN
clean_df = clean_df.dropna(axis=1, how='all')

# --- STEP B: REMOVE EMPTY ROWS ---
# Define which columns will be used to evaluate whether the row is "empty"
# (Using the current columns of clean_df)
data_columns = [col for col in clean_df.columns if col != 'SEQN']

# how="all" means that a row is removed only if ALL data columns are NaN
clean_df = clean_df.dropna(subset=data_columns, how='all')

# --- STEP C: SAVE RESULTS ---
output_file_csv = 'P_BIOPRO_clean.csv'
clean_df.to_csv(output_file_csv, index=False, encoding='utf-8-sig')

# --- CHANGE REPORT ---
print("-" * 30)
print("FINAL CLEANING REPORT")
print("-" * 30)
print(f"Original rows: {len(df)}")
print(f"Resulting rows: {len(clean_df)}")
print(f"Original columns: {len(df.columns)}")
print(f"Columns removed by list: {len(columns_to_remove)}")
print(f"Final columns: {len(clean_df.columns)}")
print(f"File saved as: {output_file_csv}")

------------------------------
FINAL CLEANING REPORT
------------------------------
Original rows: 10409
Resulting rows: 9477
Original columns: 41
Columns removed by list: 16
Final columns: 25
File saved as: P_BIOPRO_clean.csv


Initially, the dataset contained 10,409 rows and 41 columns. After the data cleaning process, the dataset was reduced to 9,477 rows and 25 columns, containing effective data suitable for further analysis without any issues. The process generates a new table named "P_BIOPRO_clean.csv". An important aspect to note is that the file extension was changed to a more accessible and user-friendly format.

In [40]:
import pandas as pd

df_p_biopro_limpio = pd.read_csv("P_BIOPRO_clean.csv")

df_p_biopro_limpio.head()

,SEQN,LBXSATSI,LBDSALSI,LBXSAPSI,LBXSASSI,LBXSC3SI,LBDSBUSI,LBXSCLSI,LBXSCK,LBDSCRSI,...,LBXSOSSI,LBDSPHSI,LBXSKSI,LBXSNASI,LBDSTBSI,LBDSCASI,LBDSCHSI,LBDSTPSI,LBDSTRSI,LBDSUASI
0,109264.0,8.0,43.0,172.0,18.0,26.0,3.21,102.0,48.0,50.39,...,276.0,1.453,4.0,139.0,8.55,2.350,4.396,69.0,0.610,208.2
1,109266.0,15.0,38.0,74.0,14.0,25.0,2.86,99.0,109.0,55.69,...,277.0,1.033,3.6,140.0,8.55,2.200,5.146,68.0,0.971,291.5
2,109271.0,8.0,38.0,88.0,14.0,30.0,2.86,101.0,85.0,68.95,...,279.0,1.292,4.3,141.0,8.55,2.225,3.827,73.0,1.140,297.4
3,109273.0,35.0,46.0,70.0,27.0,27.0,3.93,100.0,48.0,81.33,...,276.0,0.872,4.3,139.0,10.26,2.425,4.344,74.0,2.010,309.3
4,109274.0,19.0,40.0,49.0,20.0,28.0,3.93,99.0,93.0,65.42,...,278.0,0.872,3.9,138.0,10.26,2.275,2.715,68.0,1.705,303.3


Additionally, when comparing the columns of the original file with those of the new file, a considerable reduction in attributes can be observed, resulting in 25 attributes available for further analysis. The following section presents a comparison of the attributes and dimensions of the original and cleaned datasets from the "P_BIOPRO.xpt" file:

In [41]:
df_p_biopro.columns

Index(['SEQN', 'LBXSATSI', 'LBDSATLC', 'LBXSAL', 'LBDSALSI', 'LBXSAPSI',
       'LBXSASSI', 'LBXSC3SI', 'LBXSBU', 'LBDSBUSI', 'LBXSCLSI', 'LBXSCK',
       'LBXSCR', 'LBDSCRSI', 'LBXSGB', 'LBDSGBSI', 'LBXSGL', 'LBDSGLSI',
       'LBXSGTSI', 'LBDSGTLC', 'LBXSIR', 'LBDSIRSI', 'LBXSLDSI', 'LBXSOSSI',
       'LBXSPH', 'LBDSPHSI', 'LBXSKSI', 'LBXSNASI', 'LBXSTB', 'LBDSTBSI',
       'LBDSTBLC', 'LBXSCA', 'LBDSCASI', 'LBXSCH', 'LBDSCHSI', 'LBXSTP',
       'LBDSTPSI', 'LBXSTR', 'LBDSTRSI', 'LBXSUA', 'LBDSUASI'],
      dtype='str')

In [42]:
df_p_biopro_limpio.columns

Index(['SEQN', 'LBXSATSI', 'LBDSALSI', 'LBXSAPSI', 'LBXSASSI', 'LBXSC3SI',
       'LBDSBUSI', 'LBXSCLSI', 'LBXSCK', 'LBDSCRSI', 'LBDSGBSI', 'LBDSGLSI',
       'LBXSGTSI', 'LBDSIRSI', 'LBXSLDSI', 'LBXSOSSI', 'LBDSPHSI', 'LBXSKSI',
       'LBXSNASI', 'LBDSTBSI', 'LBDSCASI', 'LBDSCHSI', 'LBDSTPSI', 'LBDSTRSI',
       'LBDSUASI'],
      dtype='str')

In [43]:
df_p_biopro.shape

(10409, 41)

In [44]:
df_p_biopro_limpio.shape

(9477, 25)

# Renaming to More Readable Names

The attribute names are now changed to more readable and descriptive names to facilitate the analysis and improve the overall usability of the dataset.

In [45]:
# 1. Specific mapping dictionary for the retained columns (SI units)
mapping_names = {
    'SEQN': 'Participant_ID',
    'LBXSATSI': 'Alanine Aminotransferase (ALT) (U/L)',
    'LBDSALSI': 'Albumin, refrigerated serum (g/L)',
    'LBXSAPSI': 'Alkaline Phosphatase (ALP) (IU/L)',
    'LBXSASSI': 'Aspartate Aminotransferase (AST) (U/L)',
    'LBXSC3SI': 'Bicarbonate (mmol/L)',
    'LBDSBUSI': 'Blood Urea Nitrogen (mmol/L)',
    'LBXSCLSI': 'Chloride (mmol/L)',
    'LBXSCK': 'Creatine Phosphokinase (CPK) (IU/L)',
    'LBDSCRSI': 'Creatinine, refrigerated serum (umol/L)',
    'LBDSGBSI': 'Globulin (g/L)',
    'LBDSGLSI': 'Glucose, refrigerated serum (mmol/L)',
    'LBXSGTSI': 'Gamma Glutamyl Transferase (GGT) (IU/L)',
    'LBDSIRSI': 'Iron, refrigerated serum (umol/L)',
    'LBXSLDSI': 'Lactate Dehydrogenase (LDH) (IU/L)',
    'LBXSOSSI': 'Osmolality (mmol/Kg)',
    'LBDSPHSI': 'Phosphorus (mmol/L)',
    'LBXSKSI': 'Potassium (mmol/L)',
    'LBXSNASI': 'Sodium (mmol/L)',
    'LBDSTBSI': 'Total Bilirubin (umol/L)',
    'LBDSCASI': 'Total Calcium (mmol/L)',
    'LBDSCHSI': 'Cholesterol, refrigerated serum (mmol/L)',
    'LBDSTPSI': 'Total Protein (g/L)',
    'LBDSTRSI': 'Triglycerides, refrig serum (mmol/L)',
    'LBDSUASI': 'Uric acid (umol/L)'
}

# 2. Rename the columns in the cleaned DataFrame
df_p_biopro_final = clean_df.rename(columns=mapping_names)

# 3. Export to a new CSV file
new_file_name = 'P_BIOPRO_nombres_claros.csv'
df_p_biopro_final.to_csv(new_file_name, index=False, encoding='utf-8-sig')

print(f"✅ Process completed. The file '{new_file_name}' has been generated with readable names.")
print(f"Total processed columns: {len(df_p_biopro_final.columns)}")

✅ Process completed. The file 'P_BIOPRO_nombres_claros.csv' has been generated with readable names.
Total processed columns: 25


The newly generated dataset is displayed again for visualization and inspection.

In [46]:
import pandas as pd

df_p_biopro_nombres_claros = pd.read_csv("P_BIOPRO_nombres_claros.csv")

df_p_biopro_nombres_claros.head()

,Participant_ID,Alanine Aminotransferase (ALT) (U/L),"Albumin, refrigerated serum (g/L)",Alkaline Phosphatase (ALP) (IU/L),Aspartate Aminotransferase (AST) (U/L),Bicarbonate (mmol/L),Blood Urea Nitrogen (mmol/L),Chloride (mmol/L),Creatine Phosphokinase (CPK) (IU/L),"Creatinine, refrigerated serum (umol/L)",...,Osmolality (mmol/Kg),Phosphorus (mmol/L),Potassium (mmol/L),Sodium (mmol/L),Total Bilirubin (umol/L),Total Calcium (mmol/L),"Cholesterol, refrigerated serum (mmol/L)",Total Protein (g/L),"Triglycerides, refrig serum (mmol/L)",Uric acid (umol/L)
0,109264.0,8.0,43.0,172.0,18.0,26.0,3.21,102.0,48.0,50.39,...,276.0,1.453,4.0,139.0,8.55,2.350,4.396,69.0,0.610,208.2
1,109266.0,15.0,38.0,74.0,14.0,25.0,2.86,99.0,109.0,55.69,...,277.0,1.033,3.6,140.0,8.55,2.200,5.146,68.0,0.971,291.5
2,109271.0,8.0,38.0,88.0,14.0,30.0,2.86,101.0,85.0,68.95,...,279.0,1.292,4.3,141.0,8.55,2.225,3.827,73.0,1.140,297.4
3,109273.0,35.0,46.0,70.0,27.0,27.0,3.93,100.0,48.0,81.33,...,276.0,0.872,4.3,139.0,10.26,2.425,4.344,74.0,2.010,309.3
4,109274.0,19.0,40.0,49.0,20.0,28.0,3.93,99.0,93.0,65.42,...,278.0,0.872,3.9,138.0,10.26,2.275,2.715,68.0,1.705,303.3


This concludes the data cleaning process for the "P_BIOPRO.xpt" file.

A cleaned and more readable dataset named "P_BIOPRO_nombres_claros.csv" has been obtained. This file will be used as input for further analysis in a subsequent script.
